core/win32/callback.c

```diff
@@ -5647,6 +5647,46 @@ intercept_exception(app_state_at_intercept_t *state)
              */
             mcontext_to_context(cxt, &mcontext, false /* !set_cur_seg */);
 
+            /*
+             * i#2144 : we check if this is a single step exception
+             * where we diverted the address.
+             */
+            if (pExcptRec->ExceptionCode == EXCEPTION_SINGLE_STEP) {
+                instr_t instr;
+
+                instr_init(dcontext, &instr);
+                decode(dcontext, pExcptRec->ExceptionAddress, &instr);
+                /* Checks that exception address translate on a popf. */
+                if (instr_get_opcode(&instr) == OP_popf) {
+                    instr_reset(dcontext, &instr);
+                    decode(dcontext, faulting_pc, &instr);
+                    /* Checks that exceptionwas generated by a nop. */
+                    if (instr_get_opcode(&instr) == OP_nop) {
+                        LOG(THREAD, LOG_ASYNCH, 2,
+                            "Caught generated single step exception at "PFX"\n",
+                            pExcptRec->ExceptionAddress);
+                        /* Will continue after one byte popf. */
+                        dcontext->next_tag = mcontext.pc + 1;
+                        /* Deletes fragment starting at single step exception
+                         * if it exists.
+                         */
+                        f = fragment_lookup_bb(dcontext, dcontext->next_tag);
+                        if (f) {
+                            fragment_delete(dcontext, f, FRAGDEL_ALL);
+                        }
+                        /* Sets a field so that build_bb_ilist knows when to stop. */
+                        dcontext->forged_exception_addr = dcontext->next_tag;
```

Comment:
`forged_exception_addr` looked appropriate to me since single step should be thread-specific.